### PR TITLE
G-API: Implement WarpPerspective

### PR DESCRIPTION
**Pull request by TolyaTalamanov**
### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [ ] I agree to contribute to the project under OpenCV (BSD) License.
- [ ] To the best of my knowledge, the proposed patch is not based on a code under GPL or other license that is incompatible with OpenCV
- [ ] The PR is proposed to proper branch
- [ ] There is reference to original bug report and related work
- [ ] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [ ] The feature is well documented and sample code can be built with the project CMake

```
force_builders=Custom,Custom Win,Custom Mac
build_image:Custom=ubuntu-openvino-2019r2.0:16.04
build_image:Custom Win=openvino-2019r2.0
build_image:Custom Mac=openvino-2019r2.0

test_modules:Custom=gapi
test_modules:Custom Win=gapi
test_modules:Custom Mac=gapi

buildworker:Custom=linux-1
test_opencl:Custom=ON
test_bigdata:Custom=1
test_filter:Custom=*

build_gapi_standalone:Linux x64=ade-0.1.1f
build_gapi_standalone:Win64=ade-0.1.1f
build_gapi_standalone:Mac=ade-0.1.1f
build_gapi_standalone:Linux x64 Debug=ade-0.1.1f
build_gapi_standalone:Custom Win=ade-0.1.1f
```

### CODE REVIEW

**Comment by TolyaTalamanov** [comment]
@dmatveev

File: modules/gapi/include/opencv2/gapi/core.hpp [hunk 1]
```diff
@@ -1653,6 +1661,9 @@ number of channels as src and the depth =ddepth.
 */
 GAPI_EXPORTS GMat normalize(const GMat& src, double alpha, double beta,
                             int norm_type, int ddepth = -1);
+
+GAPI_EXPORTS GMat warpPerspective(const GMat& src, const Mat& M, const Size& dsize, int flags = cv::INTER_LINEAR,
```
Comment (alalek) [inline modules/gapi/include/opencv2/gapi/core.hpp]: documentation?

File: modules/gapi/test/common/gapi_core_tests_inl.hpp [hunk 2]
```diff
@@ -1266,6 +1266,31 @@ TEST_P(SqrtTest, AccuracyTest)
     }
 }
 
+TEST_P(WarpPerspectiveTest, AccuracyTest)
+{
+    cv::Point center{in_mat1.size() / 2};
+    cv::Mat rm = cv::getRotationMatrix2D(center, angle, scale);
+    cv::Mat transform_mat = (Mat_<double>(3,3) << rm.at<double>(0, 0), rm.at<double>(0, 1), rm.at<double>(0, 2),
+                                                  rm.at<double>(1, 0), rm.at<double>(1, 1), rm.at<double>(1, 2),
+                                                  0                  , 0                  , 1                 );
```
Comment (alalek) [inline modules/gapi/test/common/gapi_core_tests_inl.hpp]: Please avoid using of this initializer - it bloats generated binary code.
Consider using of `Matx33d` and its constructor directly.

File: modules/gapi/include/opencv2/gapi/core.hpp (hunk 1 above)
Comment (TolyaTalamanov) [reply modules/gapi/include/opencv2/gapi/core.hpp]: Fixed

File: modules/gapi/test/common/gapi_core_tests_inl.hpp (hunk 2 above)
Comment (TolyaTalamanov) [reply modules/gapi/test/common/gapi_core_tests_inl.hpp]: Fixed

**Comment by TolyaTalamanov** [comment]
@dmatveev

**Comment by alalek** [comment]
@TolyaTalamanov please squash and rebase this patch

**Comment by TolyaTalamanov** [comment]
> @TolyaTalamanov please squash and rebase this patch

@alalek Squashed

**Comment by TolyaTalamanov** [comment]
@dmatveev

File: modules/gapi/include/opencv2/gapi/core.hpp [hunk 3]
```diff
@@ -486,10 +486,18 @@ namespace core {
...
+                        "cv::gapi::warpPerspective supports only cv::BORDER_CONSTANT and cv::BORDER_REPLICATE border modes");
+            return in.withType(in.depth, in.chan).withSize(dsize);
+        }
+    };
+
     G_TYPED_KERNEL(GWarpAffine, <GMat(GMat, const Mat&, Size, int, int, const cv::Scalar&)>, "org.opencv.core.warpAffine") {
         static GMatDesc outMeta(GMatDesc in, const Mat&, Size dsize, int, int border_mode, const cv::Scalar&) {
             GAPI_Assert(border_mode != cv::BORDER_TRANSPARENT &&
-                        "cv::BORDER_TRANSPARENT mode isn't support in G-API");
+                        "cv::BORDER_TRANSPARENT mode isn't support in cv::gapi::warpAffine");
```
Comment (dmatveev) [inline modules/gapi/include/opencv2/gapi/core.hpp]: is not supported

File: modules/gapi/test/common/gapi_core_tests.hpp [hunk 4]
```diff
@@ -142,6 +142,9 @@ struct BackendOutputAllocationTest : TestWithParamBase<>
 struct BackendOutputAllocationLargeSizeWithCorrectSubmatrixTest : BackendOutputAllocationTest {};
 GAPI_TEST_FIXTURE(ReInitOutTest, initNothing, <cv::Size>, 1, out_sz)
 
+GAPI_TEST_FIXTURE(WarpPerspectiveTest, initMatrixRandU, FIXTURE_API(CompareMats, double , double, int, int, cv::Scalar), 6,
+cmpF, angle, scale, flags, border_mode, border_value)
```
Comment (dmatveev) [inline modules/gapi/test/common/gapi_core_tests.hpp]: Please think of people with vertical monitors (like me) or people who split their screens vertically.
This code can be perfectly aligned

File: modules/gapi/test/common/gapi_core_tests_inl.hpp [hunk 5]
```diff
@@ -1266,6 +1266,30 @@ TEST_P(SqrtTest, AccuracyTest)
     }
 }
 
+TEST_P(WarpPerspectiveTest, AccuracyTest)
+{
+    cv::Point center{in_mat1.size() / 2};
+    cv::Mat rm = cv::getRotationMatrix2D(center, angle, scale);
+    cv::Matx33d transform_mat(rm.at<double>(0, 0), rm.at<double>(0, 1), rm.at<double>(0, 2),
+                              rm.at<double>(1, 0), rm.at<double>(1, 1), rm.at<double>(1, 2),
+                              0                  , 0                  , 1                 );
```
Comment (dmatveev) [inline modules/gapi/test/common/gapi_core_tests_inl.hpp]: So in fact you concatenate a `{0,0,1}` vector with a 3x2 mat to get a 3x3 mat.

I believe OpenCV provides a better way to do that :)

File: modules/gapi/include/opencv2/gapi/core.hpp (hunk 3 above)
Comment (TolyaTalamanov) [reply modules/gapi/include/opencv2/gapi/core.hpp]: Done

File: modules/gapi/test/common/gapi_core_tests.hpp (hunk 4 above)
Comment (TolyaTalamanov) [reply modules/gapi/test/common/gapi_core_tests.hpp]: Done

File: modules/gapi/test/common/gapi_core_tests_inl.hpp (hunk 5 above)
Comment (TolyaTalamanov) [reply modules/gapi/test/common/gapi_core_tests_inl.hpp]: Done

**Comment by TolyaTalamanov** [comment]
@dmatveev

**Comment by TolyaTalamanov** [comment]
@dmatveev

**Comment by TolyaTalamanov** [comment]
@alalek

File: modules/gapi/test/common/gapi_core_tests_inl.hpp (hunk 5 above)
Comment (alalek) [reply modules/gapi/test/common/gapi_core_tests_inl.hpp]: Not sure that new version is simpler for review.
To reduce amount of code you may use this "overload" here:
```
Matx23d getRotationMatrix2D_(Point2f center, double angle, double scale);
```

**Pull request merged**